buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java

```diff
@@ -667,17 +667,30 @@ public ByteBuf setZero(int index, int length) {
 
     @Override
     public int setCharSequence(int index, CharSequence sequence, Charset charset) {
+        return setCharSequence0(index, sequence, charset, false);
+    }
+
+    private int setCharSequence0(int index, CharSequence sequence, Charset charset, boolean expand) {
         if (charset.equals(CharsetUtil.UTF_8)) {
-            ensureWritable(ByteBufUtil.utf8MaxBytes(sequence));
+            int length = ByteBufUtil.utf8MaxBytes(sequence);
+            if (expand) {
+                ensureWritable0(length);
+            }
+            checkIndex(index, length);
```

Comment:
Agreed. Lets resolve https://github.com/netty/netty/pull/6953#discussion_r126544205 before this PR.